File: amplify/backend/backfill-candidate-notifications.py
"""
Backfill notifications for candidates based on existing applications.
Creates notifications for:
- Application submitted successfully
- CV accepted by employer
- CV rejected by employer

Run with: python backfill-candidate-notifications.py
Requires: boto3, AWS credentials configured
"""

import boto3
import json
import uuid
from datetime import datetime, timezone

REGION = 'ap-southeast-1'

dynamodb = boto3.resource('dynamodb', region_name=REGION)
applications_table = dynamodb.Table('StandardApplications')
notifications_table = dynamodb.Table('Notifications')

def scan_all_applications():
    """Scan all applications from DynamoDB with pagination"""
    items = []
    last_key = None
    
    while True:
        scan_kwargs = {}
        if last_key:
            scan_kwargs['ExclusiveStartKey'] = last_key
        
        response = applications_table.scan(**scan_kwargs)
        items.extend(response.get('Items', []))
        
        last_key = response.get('LastEvaluatedKey')
        if not last_key:
            break
    
    return items

def create_notification(notification):
    """Create a notification in DynamoDB"""
    notification_id = f"NOTIF-{datetime.now(timezone.utc).strftime('%Y%m%d')}-{str(uuid.uuid4())[:8]}"
    
    item = {
        'notificationId': notification_id,
        'type': notification['type'],
        'title': notification['title'],
        'titleEn': notification.get('titleEn', notification['title']),
        'message': notification['message'],
        'messageEn': notification.get('messageEn', notification['message']),
        'recipientId': notification['recipientId'],
        'recipientRole': notification['recipientRole'],
        'senderId': notification.get('senderId', 'system'),
        'senderName': notification.get('senderName', 'Ốp Pờ'),
        'data': notification.get('data', {}),
        'icon': notification.get('icon', 'bell'),
        'color': notification.get('color', '#3b82f6'),
        'actionUrl': notification.get('actionUrl', ''),
        'actionText': notification.get('actionText', ''),
        'actionTextEn': notification.get('actionTextEn', ''),
        'read': True,  # Mark as read since these are old
        'deleted': False,
        'createdAt': notification['createdAt']
    }
    
    notifications_table.put_item(Item=item)
    return notification_id

def backfill():
    print("=" * 60)
    print("BACKFILL CANDIDATE NOTIFICATIONS")
    print("=" * 60)
    
    # Get all applications
    print("\n[1/3] Scanning all applications...")
    applications = scan_all_applications()
    print(f"  Found {len(applications)} applications total")
    
    # Group by candidate
    candidate_apps = {}
    for app in applications:
        candidate_id = app.get('candidateId')
        if not candidate_id or candidate_id == 'anonymous':
            continue
        if candidate_id not in candidate_apps:
            candidate_apps[candidate_id] = []
        candidate_apps[candidate_id].append(app)
    
    print(f"  Found {len(candidate_apps)} unique candidates")
    
    # Create notifications
    print("\n[2/3] Creating notifications...")
    created_count = 0
    skipped_count = 0
    
    for candidate_id, apps in candidate_apps.items():
        for app in apps:
            status = app.get('status', 'pending')
            job_title = app.get('jobTitle', 'Công việc')
            company_name = app.get('employerName', app.get('companyName', 'Nhà tuyển dụng'))
            job_id = app.get('jobId')
            job_type = app.get('jobType', 'standard')
            is_quick_job = job_type == 'quick'
            created_at = app.get('createdAt', app.get('appliedAt', datetime.now(timezone.utc).isoformat()))
            
            # 1. Application submitted notification
            try:
                create_notification({
                    'type': 'system',
                    'title': 'Ứng tuyển thành công',
                    'titleEn': 'Application submitted successfully',
                    'message': f'Bạn đã ứng tuyển thành công vào vị trí {job_title} tại {company_name}. Nhà tuyển dụng sẽ xem xét hồ sơ của bạn sớm nhất có thể.',
                    'messageEn': f'You have successfully applied for the {job_title} position at {company_name}. The employer will review your profile as soon as possible.',
                    'recipientId': candidate_id,
                    'recipientRole': 'candidate',
                    'senderId': 'system',
                    'senderName': 'Ốp Pờ',
                    'data': {
                        'jobId': job_id,
                        'jobTitle': job_title,
                        'companyName': company_name,
                        'isQuickJob': is_quick_job
                    },
                    'icon': 'briefcase',
                    'color': '#3b82f6',
                    'actionUrl': '/candidate/jobs?tab=shift' if is_quick_job else '/candidate/jobs?tab=standard',
                    'actionText': 'Xem việc đã ứng tuyển',
                    'actionTextEn': 'View applied jobs',
                    'createdAt': created_at
                })
                created_count += 1
            except Exception as e:
                print(f"  ❌ Error creating submitted notification: {e}")
                skipped_count += 1
            
            # 2. If accepted - create acceptance notification
            if status == 'accepted':
                accepted_at = app.get('acceptedAt', app.get('updatedAt', created_at))
                try:
                    create_notification({
                        'type': 'success',
                        'title': 'CV của bạn đã được chấp nhận',
                        'titleEn': 'Your CV has been accepted',
                        'message': f'CV của bạn đã được {company_name} chấp nhận cho vị trí {job_title}. Bạn có thể nhắn tin cho Nhà tuyển dụng ngay ở phần bong bóng chat bên phải dưới màn hình.',
                        'messageEn': f'Your CV was accepted by {company_name} for the {job_title} position. You can message the employer using the chat bubble at the bottom right of the screen.',
                        'recipientId': candidate_id,
                        'recipientRole': 'candidate',
                        'senderId': app.get('employerId', 'employer'),
                        'senderName': company_name,
                        'data': {
                            'jobId': job_id,
                            'jobTitle': job_title,
                            'companyName': company_name,
                            'isQuickJob': is_quick_job
                        },
                        'icon': 'check-circle',
                        'color': '#10b981',
                        'actionUrl': '/candidate/jobs?tab=shift' if is_quick_job else '/candidate/jobs?tab=standard',
                        'actionText': 'Xem việc làm',
                        'actionTextEn': 'View jobs',
                        'createdAt': accepted_at
                    })
                    created_count += 1
                except Exception as e:
                    print(f"  ❌ Error creating accepted notification: {e}")
                    skipped_count += 1
            
            # 3. If rejected - create rejection notification
            elif status == 'rejected':
                rejected_at = app.get('rejectedAt', app.get('updatedAt', created_at))
                try:
                    create_notification({
                        'type': 'system',
                        'title': 'CV của bạn chưa được chấp nhận',
                        'titleEn': 'Your CV was not accepted',
                        'message': f'CV của bạn cho vị trí {job_title} tại {company_name} chưa được chấp nhận. Bạn có thể cập nhật hồ sơ và tiếp tục ứng tuyển các công việc phù hợp khác.',
                        'messageEn': f'Your CV for the {job_title} position at {company_name} was not accepted. You can update your profile and apply for other suitable jobs.',
                        'recipientId': candidate_id,
                        'recipientRole': 'candidate',
                        'senderId': app.get('employerId', 'employer'),
                        'senderName': company_name,
                        'data': {
                            'jobId': job_id,
                            'jobTitle': job_title,
                            'companyName': company_name,
                            'isQuickJob': is_quick_job
                        },
                        'icon': 'alert-circle',
                        'color': '#ef4444',
                        'actionUrl': '/candidate/jobs?tab=shift' if is_quick_job else '/candidate/jobs?tab=standard',
                        'actionText': 'Xem việc làm khác',
                        'actionTextEn': 'Browse other jobs',
                        'createdAt': rejected_at
                    })
                    created_count += 1
                except Exception as e:
                    print(f"  ❌ Error creating rejected notification: {e}")
                    skipped_count += 1
            
            # 4. If completed - create completion notification
            elif status == 'completed':
                completed_at = app.get('completedAt', app.get('updatedAt', created_at))
                try:
                    create_notification({
                        'type': 'success',
                        'title': 'Công việc đã hoàn thành',
                        'titleEn': 'Job completed',
                        'message': f'Bạn đã hoàn thành công việc {job_title} tại {company_name}. Cảm ơn bạn đã tham gia!',
                        'messageEn': f'You have completed the {job_title} job at {company_name}. Thank you for participating!',
                        'recipientId': candidate_id,
                        'recipientRole': 'candidate',
                        'senderId': app.get('employerId', 'employer'),
                        'senderName': company_name,
                        'data': {
                            'jobId': job_id,
                            'jobTitle': job_title,
                            'companyName': company_name,
                            'isQuickJob': is_quick_job
                        },
                        'icon': 'check-circle',
                        'color': '#10b981',
                        'actionUrl': '/candidate/jobs?tab=shift' if is_quick_job else '/candidate/jobs?tab=standard',
                        'actionText': 'Xem lịch sử',
                        'actionTextEn': 'View history',
                        'createdAt': completed_at
                    })
                    created_count += 1
                except Exception as e:
                    print(f"  ❌ Error creating completed notification: {e}")
                    skipped_count += 1
    
    print(f"\n[3/3] Summary:")
    print(f"  ✅ Created: {created_count} notifications")
    print(f"  ⚠️  Skipped: {skipped_count} notifications")
    print(f"\n{'=' * 60}")
    print("BACKFILL COMPLETE!")
    print("=" * 60)

if __name__ == '__main__':
    backfill()

File: amplify/backend/backfill-candidate-notifications.js
/**
 * Backfill candidate notifications from existing applications.
 * Run with: node backfill-candidate-notifications.js
 * Requires: AWS credentials configured, aws-sdk
 */

const { DynamoDBClient, ScanCommand, PutItemCommand } = require('@aws-sdk/client-dynamodb');
const { randomUUID } = require('crypto');

const REGION = 'ap-southeast-1';
const APP_TABLE = 'StandardApplications';
const NOTIF_TABLE = 'Notifications';

const client = new DynamoDBClient({ region: REGION });

async function scanAllApplications() {
  const items = [];
  let lastKey = undefined;

  do {
    const params = {
      TableName: APP_TABLE,
      ProjectionExpression: 'candidateId, #s, jobTitle, employerName, jobId, jobType, createdAt, appliedAt, acceptedAt, rejectedAt, completedAt, updatedAt, employerId',
      ExpressionAttributeNames: { '#s': 'status' }
    };
    if (lastKey) params.ExclusiveStartKey = lastKey;

    const response = await client.send(new ScanCommand(params));
    items.push(...(response.Items || []));
    lastKey = response.LastEvaluatedKey;
  } while (lastKey);

  return items;
}

async function createNotification(notif) {
  const notifId = `NOTIF-${new Date().toISOString().slice(0,10).replace(/-/g,'')}-${randomUUID().slice(0,8)}`;
  
  const item = {
    notificationId: { S: notifId },
    type: { S: notif.type },
    title: { S: notif.title },
    titleEn: { S: notif.titleEn },
    message: { S: notif.message },
    messageEn: { S: notif.messageEn },
    recipientId: { S: notif.recipientId },
    recipientRole: { S: 'candidate' },
    senderId: { S: notif.senderId || 'system' },
    senderName: { S: notif.senderName || 'Ốp Pờ' },
    data: { M: {
      jobId: { S: notif.jobId || '' },
      jobTitle: { S: notif.jobTitle || '' },
      companyName: { S: notif.companyName || '' },
      isQuickJob: { BOOL: notif.isQuickJob || false }
    }},
    icon: { S: notif.icon },
    color: { S: notif.color },
    actionUrl: { S: notif.actionUrl },
    actionText: { S: notif.actionText },
    actionTextEn: { S: notif.actionTextEn },
    read: { BOOL: true },
    deleted: { BOOL: false },
    createdAt: { S: notif.createdAt }
  };

  await client.send(new PutItemCommand({ TableName: NOTIF_TABLE, Item: item }));
}

async function main() {
  console.log('========================================');
  console.log('BACKFILL CANDIDATE NOTIFICATIONS');
  console.log('========================================\n');

  console.log('[1/3] Scanning applications...');
  const applications = await scanAllApplications();
  console.log(`  Found ${applications.length} applications\n`);

  console.log('[2/3] Creating notifications...');
  let created = 0;
  let skipped = 0;

  for (const app of applications) {
    const candidateId = app.candidateId?.S;
    if (!candidateId || candidateId === 'anonymous') {
      skipped++;
      continue;
    }

    const status = app.status?.S || 'pending';
    const jobTitle = app.jobTitle?.S || 'Công việc';
    const companyName = app.employerName?.S || 'Nhà tuyển dụng';
    const jobId = app.jobId?.S || '';
    const jobType = app.jobType?.S || 'standard';
    const isQuickJob = jobType === 'quick';
    const createdAt = app.createdAt?.S || app.appliedAt?.S || '2026-01-01T00:00:00Z';
    const employerId = app.employerId?.S || 'employer';
    const actionUrl = isQuickJob ? '/candidate/jobs?tab=shift' : '/candidate/jobs?tab=standard';

    // 1. Application submitted notification
    try {
      await createNotification({
        type: 'system',
        title: 'Ứng tuyển thành công',
        titleEn: 'Application submitted',
        message: `Bạn đã ứng tuyển thành công vào vị trí ${jobTitle} tại ${companyName}.`,
        messageEn: `You applied for ${jobTitle} at ${companyName}.`,
        recipientId: candidateId,
        senderId: 'system',
        senderName: 'Ốp Pờ',
        jobId, jobTitle, companyName, isQuickJob,
        icon: 'briefcase',
        color: '#3b82f6',
        actionUrl,
        actionText: 'Xem việc đã ứng tuyển',
        actionTextEn: 'View applied jobs',
        createdAt
      });
      created++;
    } catch (e) {
      console.error(`  ❌ Error: ${e.message}`);
    }

    // 2. Status-based notification
    if (status === 'accepted') {
      const acceptedAt = app.acceptedAt?.S || app.updatedAt?.S || createdAt;
      try {
        await createNotification({
          type: 'success',
          title: 'CV của bạn đã được chấp nhận',
          titleEn: 'Your CV has been accepted',
          message: `CV của bạn đã được ${companyName} chấp nhận cho vị trí ${jobTitle}.`,
          messageEn: `Your CV was accepted by ${companyName} for ${jobTitle}.`,
          recipientId: candidateId,
          senderId: employerId,
          senderName: companyName,
          jobId, jobTitle, companyName, isQuickJob,
          icon: 'check-circle',
          color: '#10b981',
          actionUrl,
          actionText: 'Xem việc làm',
          actionTextEn: 'View jobs',
          createdAt: acceptedAt
        });
        created++;
      } catch (e) {
        console.error(`  ❌ Error: ${e.message}`);
      }
    } else if (status === 'rejected') {
      const rejectedAt = app.rejectedAt?.S || app.updatedAt?.S || createdAt;
      try {
        await createNotification({
          type: 'system',
          title: 'CV của bạn chưa được chấp nhận',
          titleEn: 'Your CV was not accepted',
          message: `CV của bạn cho vị trí ${jobTitle} tại ${companyName} chưa được chấp nhận.`,
          messageEn: `Your CV for ${jobTitle} at ${companyName} was not accepted.`,
          recipientId: candidateId,
          senderId: employerId,
          senderName: companyName,
          jobId, jobTitle, companyName, isQuickJob,
          icon: 'alert-circle',
          color: '#ef4444',
          actionUrl,
          actionText: 'Xem việc làm khác',
          actionTextEn: 'Browse other jobs',
          createdAt: rejectedAt
        });
        created++;
      } catch (e) {
        console.error(`  ❌ Error: ${e.message}`);
      }
    } else if (status === 'completed') {
      const completedAt = app.completedAt?.S || app.updatedAt?.S || createdAt;
      try {
        await createNotification({
          type: 'success',
          title: 'Công việc đã hoàn thành',
          titleEn: 'Job completed',
          message: `Bạn đã hoàn thành công việc ${jobTitle} tại ${companyName}. Cảm ơn bạn!`,
          messageEn: `You completed ${jobTitle} at ${companyName}. Thank you!`,
          recipientId: candidateId,
          senderId: employerId,
          senderName: companyName,
          jobId, jobTitle, companyName, isQuickJob,
          icon: 'check-circle',
          color: '#10b981',
          actionUrl,
          actionText: 'Xem lịch sử',
          actionTextEn: 'View history',
          createdAt: completedAt
        });
        created++;
      } catch (e) {
        console.error(`  ❌ Error: ${e.message}`);
      }
    }

    if (created % 10 === 0 && created > 0) {
      process.stdout.write(`  ... ${created} created\r`);
    }
  }

  console.log(`\n\n[3/3] Summary:`);
  console.log(`  ✅ Created: ${created} notifications`);
  console.log(`  ⚠️  Skipped: ${skipped} (no candidateId)`);
  console.log('\n========================================');
  console.log('BACKFILL COMPLETE!');
  console.log('========================================');
}

main().catch(e => {
  console.error('Fatal error:', e);
  process.exit(1);
});
